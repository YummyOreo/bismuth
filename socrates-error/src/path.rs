#![allow(unused)]
use std::{io::stdout, path::PathBuf};

use crossterm::{
    cursor,
    event::{read, Event, KeyCode},
    execute,
    style::{Color, Print, ResetColor, SetForegroundColor},
    terminal, Result,
};

use crate::error_ui;

fn update_input(path: &mut String) -> Option<()> {
    if let Event::Key(k) = read().expect("Should be able to read") {
        match k.code {
            KeyCode::Char(x) => {
                path.push(x);
                execute!(stdout(), Print(x.to_string()));
            }
            KeyCode::Backspace => {
                path.pop();
                let print_path = format!("Enter new path: {}", path);
                execute!(
                    stdout(),
                    terminal::Clear(terminal::ClearType::CurrentLine),
                    cursor::MoveToColumn(0),
                    Print(print_path),
                );
            }
            KeyCode::Enter => return Some(()),
            _ => {}
        }
    }
    None
}

pub fn md_file_error(_description: &str) -> Option<PathBuf> {
    let options = vec![
        // "Make folder at path.".to_string(),
        "Chose another path".to_string(),
    ];

    let option = error_ui(&options)?;

    match option {
        // 1 => {}
        1 => {
            execute!(
                stdout(),
                cursor::MoveUp(u16::try_from(options.len()).unwrap() + 1_u16),
                terminal::Clear(terminal::ClearType::CurrentLine),
                cursor::SavePosition
            );
            for i in 0..options.len() + 1 {
                execute!(
                    stdout(),
                    cursor::MoveDown(1),
                    terminal::Clear(terminal::ClearType::CurrentLine),
                    cursor::RestorePosition,
                );
            }

            execute!(
                stdout(),
                ResetColor,
                Print("Enter new path: "),
                cursor::Show
            );
            let mut path = "".to_string();
            loop {
                if update_input(&mut path).is_some() {
                    break;
                }
            }
            return Some(PathBuf::from(path));
        }
        _ => panic!("Something has gone terribly wrong"),
    }
    None
}
