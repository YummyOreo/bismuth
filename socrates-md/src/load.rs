use std::fs;
use std::path::PathBuf;

use crate::{MarkdownFile, MarkdownFileError};

pub fn load_from_dir(path: &PathBuf) -> Result<Vec<MarkdownFile>, MarkdownFileError> {
    let mut files = vec![];
    if !path.is_dir() {
        return Err(MarkdownFileError::NotDirectoryError(
            path.to_string_lossy().to_string(),
        ));
    }

    let paths = fs::read_dir(path).expect("Should be directory");
    for file in paths {
        let file_path = file.unwrap().path();
        let rel = file_path
            .to_string_lossy()
            .replace(&path.to_string_lossy().to_string(), ".");
        let file_rel = PathBuf::from(rel);
        if file_path.is_dir() {
            if let Ok(mut m) = load_from_dir(&file_path) {
                files.append(&mut m)
            }
        } else if file_path.is_file() {
            if let Ok(m) = MarkdownFile::load_file(&file_path, &file_rel) {
                files.push(m)
            }
        }
    }

    Ok(files)
}

// # Tests:
//
// TODO: Fix this, broken because:
// Could load in any way order, should account for this

#[cfg(test)]
mod test {
    use super::*;
    use std::path::PathBuf;

    fn snapshot(path: &str) -> String {
        let path = PathBuf::from(path).canonicalize().unwrap();
        let mut files = load_from_dir(&path).unwrap();
        for mut file in &mut files {
            let new_path = file.path.to_string_lossy().replace('\\', "/").to_lowercase();
            file.path = PathBuf::from(new_path);
        }
        files.sort_by(|a,b| a.path.cmp(&b.path));
        format!("{files:#?}")
    }

    macro_rules! snapshot {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("../testdata/output/");
                settings.bind(|| {
                    insta::assert_snapshot!(snapshot($path));
                });
            }
        };
    }

    snapshot!(test_load, "./testdata/tests/");
}
