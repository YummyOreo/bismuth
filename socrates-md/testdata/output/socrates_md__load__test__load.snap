---
source: socrates-md/src/load.rs
expression: "snapshot(\"./testdata/tests/\")"
---
[
    MarkdownFile {
        content: "# this is test 0\r\n",
        path: "./test.markdown",
    },
    MarkdownFile {
        content: "# this is test 1\r\n",
        path: "./test1.md",
    },
]
